### PR TITLE
[PR1/5] Add S3 Storage type and S3StorageClient

### PR DESCRIPTION
**Pull request by jainlavina**
## Summary
Openhouse catalog currently supports HDFS as the storage backend. The end goal of this effort is to add the integration with S3 so that the storage backend can be configured to be S3 vs HDFS based on storage type.
The entire work will be done via a series of PRs:
1. Add S3 Storage type and S3StorageClient.
2. Add base class for StorageClient and move common logic like validation of properties there to avoid code duplication.
3. Add S3Storage implementation that uses S3StorageClient.
4. Add support for using S3FileIO for S3 storage type.
5. Add a recipe for end-to-end testing in docker.

This PR addresses 1 by adding S3 storage type and S3StorageClient.

## Changes

- [ ] Client-facing API Changes
- [ ] Internal API Changes
- [ ] Bug Fixes
- [x] New Features
- [ ] Performance Improvements
- [ ] Code Style
- [ ] Refactoring
- [ ] Documentation
- [ ] Tests

For all the boxes checked, please include additional details of the changes made in this pull request.  

## Testing Done
Added unit tests. This is one of the first few PRs to complete plugging in S3 storage. More test recipes to test with S3 storage will be added when S3 storage is completely plugged in. Currently tested oh-hadoop-spark recipe to ensure that  the HDFS integration is not broken by this change.

- [x] Manually Tested on local docker setup. Please include commands ran, and their output.
- [ ] Added new tests for the changes made.
- [ ] Updated existing tests to reflect the changes made.
- [ ] No tests added or updated. Please explain why. If unsure, please feel free to ask for help.
- [ ] Some other form of testing like staging or soak time in production. Please explain.

For all the boxes checked, include a detailed description of the testing done for the changes made in this pull request.

Docker testing oh-hadoop-spark recipe:

1. Create table:

$ curl "${curlArgs[@]}" -XPOST http://localhost:8000/v1/databases/d3/tables/ \
> --data-raw '{
>   "tableId": "t1",
>   "databaseId": "d3",
>   "baseTableVersion": "INITIAL_VERSION",
>   "clusterId": "LocalHadoopCluster",
>   "schema": "{\"type\": \"struct\", \"fields\": [{\"id\": 1,\"required\": true,\"name\": \"id\",\"type\": \"string\"},{\"id\": 2,\"required\": true,\"name\": \"name\",\"type\": \"string\"},{\"id\": 3,\"required\": true,\"name\": \"ts\",\"type\": \"timestamp\"}]}",
>   "timePartitioning": {
>     "columnName": "ts",
>     "granularity": "HOUR"
>   },
>   "clustering": [
>     {
>       "columnName": "name"
>     }
>   ],
>   "tableProperties": {
>     "key": "value"
>   }
> }' | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2174    0  1600  100   574    548    196  0:00:02  0:00:02 --:--:--   745
{
   "clusterId" : "LocalHadoopCluster",
   "clustering" : [
      {
         "columnName" : "name",
         "transform" : null
      }
   ],
   "creationTime" : 1717098450027,
   "databaseId" : "d3",
   "lastModifiedTime" : 1717098450027,
   "policies" : null,
   "schema" : "{\"type\":\"struct\",\"schema-id\":0,\"fields\":[{\"id\":1,\"name\":\"id\",\"required\":true,\"type\":\"string\"},{\"id\":2,\"name\":\"name\",\"required\":true,\"type\":\"string\"},{\"id\":3,\"name\":\"ts\",\"required\":true,\"type\":\"timestamp\"}]}",
   "tableCreator" : "DUMMY_ANONYMOUS_USER",
   "tableId" : "t1",
   "tableLocation" : "hdfs://namenode:9000/data/openhouse/d3/t1-30f90df6-4c45-47ee-916c-7cf0bdea5d4d/00000-5fec4fc6-4bf9-4bab-b6a9-112967a57497.metadata.json",
   "tableProperties" : {
      "key" : "value",
      "openhouse.clusterId" : "LocalHadoopCluster",
      "openhouse.creationTime" : "1717098450027",
      "openhouse.databaseId" : "d3",
      "openhouse.lastModifiedTime" : "1717098450027",
      "openhouse.tableCreator" : "DUMMY_ANONYMOUS_USER",
      "openhouse.tableId" : "t1",
      "openhouse.tableLocation" : "/data/openhouse/d3/t1-30f90df6-4c45-47ee-916c-7cf0bdea5d4d/00000-5fec4fc6-4bf9-4bab-b6a9-112967a57497.metadata.json",
      "openhouse.tableType" : "PRIMARY_TABLE",
      "openhouse.tableUUID" : "30f90df6-4c45-47ee-916c-7cf0bdea5d4d",
      "openhouse.tableUri" : "LocalHadoopCluster.d3.t1",
      "openhouse.tableVersion" : "INITIAL_VERSION",
      "policies" : "",
      "write.format.default" : "orc",
      "write.metadata.delete-after-commit.enabled" : "true",
      "write.metadata.previous-versions-max" : "28"
   },
   "tableType" : "PRIMARY_TABLE",
   "tableUUID" : "30f90df6-4c45-47ee-916c-7cf0bdea5d4d",
   "tableUri" : "LocalHadoopCluster.d3.t1",
   "tableVersion" : "INITIAL_VERSION",
   "timePartitioning" : {
      "columnName" : "ts",
      "granularity" : "HOUR"
   }
}

2. Read table:
$ curl "${curlArgs[@]}" -XGET http://localhost:8000/v1/databases/d3/tables/t1 | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1600    0  1600    0     0  22736      0 --:--:-- --:--:-- --:--:-- 22857
{
   "clusterId" : "LocalHadoopCluster",
   "clustering" : [
      {
         "columnName" : "name",
         "transform" : null
      }
   ],
   "creationTime" : 1717098450027,
   "databaseId" : "d3",
   "lastModifiedTime" : 1717098450027,
   "policies" : null,
   "schema" : "{\"type\":\"struct\",\"schema-id\":0,\"fields\":[{\"id\":1,\"name\":\"id\",\"required\":true,\"type\":\"string\"},{\"id\":2,\"name\":\"name\",\"required\":true,\"type\":\"string\"},{\"id\":3,\"name\":\"ts\",\"required\":true,\"type\":\"timestamp\"}]}",
   "tableCreator" : "DUMMY_ANONYMOUS_USER",
   "tableId" : "t1",
   "tableLocation" : "hdfs://namenode:9000/data/openhouse/d3/t1-30f90df6-4c45-47ee-916c-7cf0bdea5d4d/00000-5fec4fc6-4bf9-4bab-b6a9-112967a57497.metadata.json",
   "tableProperties" : {
      "key" : "value",
      "openhouse.clusterId" : "LocalHadoopCluster",
      "openhouse.creationTime" : "1717098450027",
      "openhouse.databaseId" : "d3",
      "openhouse.lastModifiedTime" : "1717098450027",
      "openhouse.tableCreator" : "DUMMY_ANONYMOUS_USER",
      "openhouse.tableId" : "t1",
      "openhouse.tableLocation" : "/data/openhouse/d3/t1-30f90df6-4c45-47ee-916c-7cf0bdea5d4d/00000-5fec4fc6-4bf9-4bab-b6a9-112967a57497.metadata.json",
      "openhouse.tableType" : "PRIMARY_TABLE",
      "openhouse.tableUUID" : "30f90df6-4c45-47ee-916c-7cf0bdea5d4d",
      "openhouse.tableUri" : "LocalHadoopCluster.d3.t1",
      "openhouse.tableVersion" : "INITIAL_VERSION",
      "policies" : "",
      "write.format.default" : "orc",
      "write.metadata.delete-after-commit.enabled" : "true",
      "write.metadata.previous-versions-max" : "28"
   },
   "tableType" : "PRIMARY_TABLE",
   "tableUUID" : "30f90df6-4c45-47ee-916c-7cf0bdea5d4d",
   "tableUri" : "LocalHadoopCluster.d3.t1",
   "tableVersion" : "INITIAL_VERSION",
   "timePartitioning" : {
      "columnName" : "ts",
      "granularity" : "HOUR"
   }
}

3. List tables:
$ curl "${curlArgs[@]}" -XGET http://localhost:8000/v1/databases/d3/tables/ | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1614    0  1614    0     0  30447      0 --:--:-- --:--:-- --:--:-- 31038
{
   "results" : [
      {
         "clusterId" : "LocalHadoopCluster",
         "clustering" : [
            {
               "columnName" : "name",
               "transform" : null
            }
         ],
         "creationTime" : 1717098450027,
         "databaseId" : "d3",
         "lastModifiedTime" : 1717098450027,
         "policies" : null,
         "schema" : "{\"type\":\"struct\",\"schema-id\":0,\"fields\":[{\"id\":1,\"name\":\"id\",\"required\":true,\"type\":\"string\"},{\"id\":2,\"name\":\"name\",\"required\":true,\"type\":\"string\"},{\"id\":3,\"name\":\"ts\",\"required\":true,\"type\":\"timestamp\"}]}",
         "tableCreator" : "DUMMY_ANONYMOUS_USER",
         "tableId" : "t1",
         "tableLocation" : "hdfs://namenode:9000/data/openhouse/d3/t1-30f90df6-4c45-47ee-916c-7cf0bdea5d4d/00000-5fec4fc6-4bf9-4bab-b6a9-112967a57497.metadata.json",
         "tableProperties" : {
            "key" : "value",
            "openhouse.clusterId" : "LocalHadoopCluster",
            "openhouse.creationTime" : "1717098450027",
            "openhouse.databaseId" : "d3",
            "openhouse.lastModifiedTime" : "1717098450027",
            "openhouse.tableCreator" : "DUMMY_ANONYMOUS_USER",
            "openhouse.tableId" : "t1",
            "openhouse.tableLocation" : "/data/openhouse/d3/t1-30f90df6-4c45-47ee-916c-7cf0bdea5d4d/00000-5fec4fc6-4bf9-4bab-b6a9-112967a57497.metadata.json",
            "openhouse.tableType" : "PRIMARY_TABLE",
            "openhouse.tableUUID" : "30f90df6-4c45-47ee-916c-7cf0bdea5d4d",
            "openhouse.tableUri" : "LocalHadoopCluster.d3.t1",
            "openhouse.tableVersion" : "INITIAL_VERSION",
            "policies" : "",
            "write.format.default" : "orc",
            "write.metadata.delete-after-commit.enabled" : "true",
            "write.metadata.previous-versions-max" : "28"
         },
         "tableType" : "PRIMARY_TABLE",
         "tableUUID" : "30f90df6-4c45-47ee-916c-7cf0bdea5d4d",
         "tableUri" : "LocalHadoopCluster.d3.t1",
         "tableVersion" : "INITIAL_VERSION",
         "timePartitioning" : {
            "columnName" : "ts",
            "granularity" : "HOUR"
         }
      }
   ]
}

4. Delete table:
$ curl "${curlArgs[@]}" -XDELETE http://localhost:8000/v1/databases/d3/tables/t1
lajain-mn2:oh-hadoop-spark lajain$ curl "${curlArgs[@]}" -XGET http://localhost:8000/v1/databases/d3/tables/ | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    14    0    14    0     0    331      0 --:--:-- --:--:-- --:--:--   333
{
   "results" : []
}

Testing via spark shell:
scala> spark.sql("CREATE TABLE openhouse.db.tb (ts timestamp, col1 string, col2 string) PARTITIONED BY (days(ts))").show()
++
||
++
++

scala> spark.sql("DESCRIBE TABLE openhouse.db.tb").show()
+--------------+---------+-------+
|      col_name|data_type|comment|
+--------------+---------+-------+
|            ts|timestamp|       |
|          col1|   string|       |
|          col2|   string|       |
|              |         |       |
|# Partitioning|         |       |
|        Part 0| days(ts)|       |
+--------------+---------+-------+

scala> spark.sql("INSERT INTO TABLE openhouse.db.tb VALUES (current_timestamp(), 'val1', 'val2')")
res2: org.apache.spark.sql.DataFrame = []                                       

scala> spark.sql("INSERT INTO TABLE openhouse.db.tb VALUES (date_sub(CAST(current_timestamp() as DATE), 30), 'val1', 'val2')")
res3: org.apache.spark.sql.DataFrame = []

scala> spark.sql("INSERT INTO TABLE openhouse.db.tb VALUES (date_sub(CAST(current_timestamp() as DATE), 60), 'val1', 'val2')")
res4: org.apache.spark.sql.DataFrame = []                                       

scala> spark.sql("SELECT * FROM openhouse.db.tb").show()
+--------------------+----+----+                                                
|                  ts|col1|col2|
+--------------------+----+----+
|2024-05-30 19:52:...|val1|val2|
| 2024-03-31 00:00:00|val1|val2|
| 2024-04-30 00:00:00|val1|val2|
+--------------------+----+----+

scala> spark.sql("SHOW TABLES IN openhouse.db").show()
+---------+---------+
|namespace|tableName|
+---------+---------+
|       db|       tb|
+---------+---------+

# Additional Information

- [ ] Breaking Changes
- [ ] Deprecations
- [x] Large PR broken into smaller PRs, and PR plan linked in the description.

For all the boxes checked, include additional details of the changes made in this pull request.

### CODE REVIEW

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java
Comment (HotSushi) [inline cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java:25]: Can we add validations here and possible move such common logic to `BaseStorageClient`?

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java
Comment (HotSushi) [inline cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java:15]: Do we need some javadoc for these classes?

**Review by HotSushi — COMMENTED** [review]
Thanks for these changes, generally looking good, left few comments.

**Comment by sumedhsakdeo** [comment]
Overall code looks great on first pass. 

>>  Manually Tested on local docker setup. Please include commands ran, and their output.

Was this tested on local docker setup with s3 container, or current HDFS container. 

>> Added unit tests. This is one of the first few PRs to complete plugging in S3 storage. More test recipes will be added when S3 storage is completely plugged in.

Can you update the description with a high level PR plan on what to expect after this lands? 

Also super nit, PR title [PR 1/n] Add S3 Storage type and S3StorageClient

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java
Comment (sumedhsakdeo) [reply cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java:25]: I believe S3 will have its own validations. What is the expectation for s3/aws access token here? We cannot keep it in plain-text in cluster spec, it would need to be injected into memory out of band.

Comment (sumedhsakdeo) [reply cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java:25]: Let me take that back, we could securely rebuild cluster spec in the init container. For HDFS we fetch the delegation token in the init container and write them to a fixed location, so something similar to that. Caspian will need similar functionality but we can solve it when we get there.

File: cluster/storage/build.gradle
Comment (jainlavina) [reply cluster/storage/build.gradle]: Done

**Review by sumedhsakdeo — DISMISSED** [review]
💯 please update the description with PR plan for posterity

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java
Comment (jainlavina) [reply cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java:31]: @HotSushi IS it ok to address this in the next PR? I have added a TODO.

Comment (HotSushi) [reply cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java:31]: Yes SG!

**Review by HotSushi — DISMISSED** [review]
LGTM after the javadoc ✅

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java
Comment (jainlavina) [reply cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3StorageClient.java:18]: Added javadoc comments.

File: .github/workflows/build-tag-publish.yml
Comment (HotSushi) [inline .github/workflows/build-tag-publish.yml:18]: can this change be pursued independently?

Comment (jainlavina) [reply .github/workflows/build-tag-publish.yml]: Yes.

**Review by HotSushi — APPROVED** [review]

**Pull request merged**
